File: javalib/src/main/java/com/strumenta/kolasu/javalib/ParseTreeToASTTransformer.java
package com.strumenta.kolasu.javalib;

import com.strumenta.kolasu.model.Node;
import com.strumenta.kolasu.model.Source;
import com.strumenta.kolasu.transformation.ASTTransformer;
import com.strumenta.kolasu.transformation.NodeFactory;
import com.strumenta.kolasu.validation.Issue;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import static kotlin.jvm.JvmClassMappingKt.getKotlinClass;

public class ParseTreeToASTTransformer extends com.strumenta.kolasu.mapping.ParseTreeToASTTransformer {
    public ParseTreeToASTTransformer(@NotNull List<Issue> issues, boolean allowGenericNode, @Nullable Source source, boolean throwOnUnmappedNode) {
        super(issues, allowGenericNode, source, throwOnUnmappedNode);
    }

    public ParseTreeToASTTransformer(@NotNull List<Issue> issues, boolean allowGenericNode, @Nullable Source source) {
        super(issues, allowGenericNode, source);
    }

    public ParseTreeToASTTransformer(@NotNull List<Issue> issues, boolean allowGenericNode) {
        super(issues, allowGenericNode);
    }

    public ParseTreeToASTTransformer(@NotNull List<Issue> issues) {
        super(issues);
    }

    protected <S, T extends Node> @NotNull NodeFactory<S, T> registerNodeFactory(Class<S> source, Class<T> target) {
        return registerNodeFactory(source, target, target.getName());
    }

    protected <S, T extends Node> @NotNull NodeFactory<S, T> registerNodeFactory(
            Class<S> source, Class<T> target, String nodeType
    ) {
        return registerNodeFactory(getKotlinClass(source), getKotlinClass(target), nodeType);
    }

    protected <S, T extends Node> NodeFactory<S, T> registerNodeFactory(Class<S> source, Function1<S, T> function) {
        return registerNodeFactory(getKotlinClass(source), (s, t) -> function.invoke(s));
    }

    protected <S, T extends Node> NodeFactory<S, T> registerNodeFactory(Class<S> source, Function2<S, ? super ASTTransformer, T> function) {
        return registerNodeFactory(getKotlinClass(source), function);
    }

}

File: antlr4j/src/main/kotlin/com/strumenta/kolasu/antlr4j/mapping/ParseTreeToASTTransformer.kt
package com.strumenta.kolasu.antlr4j.mapping

import com.strumenta.kolasu.antlr4j.parsing.ParseTreeOrigin
import com.strumenta.kolasu.antlr4j.parsing.withParseTreeNode
import com.strumenta.kolasu.model.NodeLike
import com.strumenta.kolasu.model.Origin
import com.strumenta.kolasu.model.Source
import com.strumenta.kolasu.transformation.ASTTransformer
import com.strumenta.kolasu.transformation.NodeTransformer
import com.strumenta.kolasu.validation.Issue
import org.antlr.v4.runtime.ParserRuleContext
import org.antlr.v4.runtime.tree.ParseTree
import kotlin.reflect.KClass

/**
 * Implements a transformation from an ANTLR parse tree (the output of the parser) to an AST (a higher-level
 * representation of the source code).
 */
open class ParseTreeToASTTransformer(
    issues: MutableList<Issue> = mutableListOf(),
    allowGenericNode: Boolean = true,
    val source: Source? = null,
    throwOnUnmappedNode: Boolean = true,
) : ASTTransformer(issues, allowGenericNode, throwOnUnmappedNode) {
    /**
     * Performs the transformation of a node and, recursively, its descendants. In addition to the overridden method,
     * it also assigns the parseTreeNode to the AST node so that it can keep track of its range.
     * However, a node transformer can override the parseTreeNode of the nodes it creates (but not the parent).
     */
    override fun transformIntoNodes(
        source: Any?,
        parent: NodeLike?,
        expectedType: KClass<out NodeLike>,
    ): List<NodeLike> {
        val transformed = super.transformIntoNodes(source, parent, expectedType)
        return transformed
            .map { node ->
                if (source is ParserRuleContext) {
                    if (node.origin == null) {
                        node.withParseTreeNode(source, this.source)
                    } else if (node.range != null && node.source == null) {
                        node.range!!.source = this.source
                    }
                }
                return listOf(node)
            }.flatten()
    }

    override fun getSource(
        node: NodeLike,
        source: Any,
    ): Any {
        val origin = node.origin
        return if (origin is ParseTreeOrigin) origin.parseTree else source
    }

    override fun asOrigin(source: Any): Origin? = if (source is ParseTree) ParseTreeOrigin(source) else null

    /**
     * Often in ANTLR grammar we have rules which wraps other rules and act as
     * wrapper. When there is only a ParserRuleContext child we can transform
     * that child and return that result.
     */
    fun <P : ParserRuleContext> registerNodeTransformerUnwrappingChild(
        kclass: KClass<P>,
    ): NodeTransformer<P, NodeLike> =
        registerNodeTransformer(kclass) { source, transformer, _ ->
            val nodeChildren = source.children.filterIsInstance<ParserRuleContext>()
            require(nodeChildren.size == 1) {
                "Node $source (${source.javaClass}) has ${nodeChildren.size} " +
                    "node children: $nodeChildren"
            }
            transformer.transform(nodeChildren[0]) as NodeLike
        }

    /**
     * Alternative to registerNodeFactoryUnwrappingChild(KClass) which is slightly more concise.
     */
    inline fun <reified P : ParserRuleContext> registerNodeFactoryUnwrappingChild(): NodeTransformer<P, NodeLike> =
        registerNodeTransformerUnwrappingChild(P::class)
}
